### ISSUE TITLE
FR: add "signed-off-by" line

### ISSUE THREAD

**Issue by yshui** [jj-vcs/jj#1399]
## Description

Support adding a `Signed-off-by: Name <Email>` line at the end of the commit message, like `git commit -s`.

Maybe even make this configurable footer.

**Comment by joyously** [comment]
Ooo, I like the alias since it's versatile.
Somewhat related is a plugin that does `fast-import` I've seen. I don't know if the plugin put the SVN revision number into the Git description, or what, but it was very useful for back-tracking (and also for gauging where you are in the history since they are sequential in SVN). 
I'm saying it could be useful to be able to add to the description from a hook, such as import.

**Comment by omasanori** [comment]
The alias is so helpful to me, thanks a lot! I use a modified version below, as I use different `user.email` for different projects.

```sh
#!/bin/sh

echo "Signed-off-by: $(jj config get user.name) <$(jj config get user.email)>" >> "$1"
```

> I've only heard of them being used in the Git project. Oh, the Linux kernel uses them too, of course.

> My feeling is that they're too specific to a few particular projects to deserve special treatment. I'm curious what others think.

For reference, DCO has been adopted by [Arm Trusted Firmware-A](https://github.com/ARM-software/arm-trusted-firmware/blob/43a6544f015b8d91036e4eb7e422a545a8f2396e/docs/process/commit-style.rst), [Chef](https://github.com/chef/chef/blob/dde8a6c8a4cce1a30e9dd627fe6fd9796a893e10/CONTRIBUTING.md), [Collabora Online](https://github.com/CollaboraOnline/online/blob/ec67c92264775b2fd52de32d2abe2f053c53be15/CONTRIBUTING.md), [Coreboot](https://github.com/coreboot/coreboot/blob/87f08bea11ecabc721e80d553ff8173c77ae19ab/Documentation/contributing/gerrit_guidelines.md), [FreeBSD](https://cgit.freebsd.org/src/tree/CONTRIBUTING.md#n40), [GCC](https://gcc.gnu.org/dco.html), [ISC Kea](https://github.com/isc-projects/kea/blob/3e7dcb311994b0b2538cf4c4d82339bfbb6b7b2a/CONTRIBUTING.md), [Lustre](https://wiki.lustre.org/Submitting_Changes), [Node.js](https://github.com/nodejs/node/blob/48345d0f6212d0c21922b5119c43c035013b7c5e/CONTRIBUTING.md), [OpenUI5](https://github.com/SAP/openui5/blob/6fea5637225124d73cb49302fb74ecf8190b695f/CONTRIBUTING.md), [U-boot](https://github.com/u-boot/u-boot/blob/c98c401dfb485b39c7453a4147b17cd4b8d10c67/doc/develop/sending_patches.rst), [Xilinx Runtime](https://github.com/Xilinx/XRT/blob/64f9555c3f217ad47b3347fb41ce76d6b8cd27ff/CONTRIBUTING.rst), among others.

**Comment by martinvonz** [comment]
Thanks for the pointers. That's quite a few large projects. I still don't really like the idea of commands and flags for such specialized features. Maybe a new footer becomes more popular in a few years. Do we add another flag for that then? Or if the "Signed-off-by:" footer falls out of fashion, then we're stuck with a useless flag (depending on what our compatibility policy is then). So maybe we instead have a `--add-signature/-s` that adds a configurable footer instead. What do you (and others) think?

**Comment by omasanori** [comment]
> I still don't really like the idea of commands and flags for such specialized features. Maybe a new footer becomes more popular in a few years. Do we add another flag for that then? Or if the "Signed-off-by:" footer falls out of fashion, then we're stuck with a useless flag (depending on what our compatibility policy is then).

Agreed. GitHub has a [support for `Co-authored-by:`](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) and Linux kernel uses [`Acked-by:`, `Co-developed-by:`, `Reported-by:,` `Tested-by:`, `Reviewed-by:`, and `Suggested-by:`](https://docs.kernel.org/process/submitting-patches.html). Configurable footer is clearly better to me.

**Comment by martinvonz** [comment]
What do you think the interface for this feature should be? A new command? A flag on `commit` (but maybe you want to add the signature to an existing commit) or `describe`? If we put it on `describe`, then the user will probably usually also want `--no-edit`, but making it implied also seems surprising. I suppose we could have both a `add-signature` command an a `describe --add-signature` flag?

**Comment by omasanori** [comment]
How about making the commit message boilerplate configurable with a template? Adding a line afterward can be implementable with an alias as you've shown.

**Comment by martinvonz** [comment]
That sounds like #1354. To make sure I understand correctly, what would you put in your template?

**Comment by omasanori** [comment]
In my case,
```
Signed-off-by: (the value of user.name) <(the value of user.email)>
```
but interpolation of configuration values is optional if a template can be set per-repo. (Inconvenient a bit, though.)

**Comment by martinvonz** [comment]
So it's mostly #1354 that you'd need then, right? Do you feel like working on that? We'd appreciate it :)

**Comment by omasanori** [comment]
Aha, yes. Not a promise but I will try. Thanks for suggestion!

**Comment by avamsi** [comment]
FYI: with #2062, you could add something like `ui.default-description = "\n\nSigned-off-by: $name <$email>"` to your config (it's just a string for now, so you'll have to repeat the values of name and email).

**Comment by samhh** [comment]
Chiming in to say that #2062 wouldn't satisfyingly cover my use case which is to sometimes but not always append "Co-authored-by" to commit messages for GitHub repos, typically in response to PR feedback or during pairing.

I currently (only just discovered Jujutsu) use a Git alias I wrote for this: https://github.com/samhh/dotfiles/blob/fad290da04f4d4f852ae2d85d65104809a26758d/home/vcs.nix#L61

The main part is:

```sh
git commit --amend --only -m "$(git log --format=%B -n1)" -m "$msg"
```

If I may dream for a moment, the prefix would be configurable for reuse:

```toml
[attributions]
co = "Co-authored-by"
```

And then somehow Jujutsu would figure out a neat way to let you select people like my script does, since I'm typically picking one or two of the same half a dozen people. Not sure how that interops with the signoff use case, mind.

Another way of looking at this is through pairing. It seems like we're conflating "commit author" and "machine user".

**Comment by ErichDonGubler** [comment]
One thing I wanted to add: If we're concerned about overly specialized CLI flags for specific trailers, we might consider a design like the [`-t/--trailer` option in the `git commit` interface](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---trailerlttokengtltvaluegt), which allows both (1) configuration of behavior when a "duplicate" trailer is specified and (2) avoids the question of "what if a new [better] option comes up", but only by taking on the complexity of something more flexible.

**Comment by dancrossnyc** [comment]
It strikes me that if one had the ability to execute some sort of command to generate a commit message template, then that, with an alias, would address most of these use-cases.

I suppose that's essentially what over-riding the editor does, as in the earliest responses.

**Comment by gmorenz** [comment]
> Out of curiosity, which project do you work on that uses such footers?

If you're still curious (or anyone else is), just ran into this on https://github.com/servo/mozjs

**Comment by ErichDonGubler** [comment]
> If you're still curious (or anyone else is), just ran into this on [servo/mozjs](https://github.com/servo/mozjs?rgh-link-date=2025-03-02T23%3A55%3A19.000Z)

AFAIK, the entire Servo project does this: https://github.com/servo/servo

**Comment by yuja** [comment]
Should be implemented by https://github.com/jj-vcs/jj/pull/6088
